rabbitmq/resource_topic_permissions.go

```diff
@@ -0,0 +1,211 @@
+)
+
+func resourceTopicPermissions() *schema.Resource {
+	return &schema.Resource{
+		Create: CreateTopicPermissions,
+		Update: UpdateTopicPermissions,
+		Read:   ReadTopicPermissions,
+		Delete: DeleteTopicPermissions,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"user": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"vhost": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "/",
+				ForceNew: true,
+			},
+
+			"permissions": {
+				Type:     schema.TypeList,
+				Required: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"exchange": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+
+						"write": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+
+						"read": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func CreateTopicPermissions(d *schema.ResourceData, meta interface{}) error {
+	rmqc := meta.(*rabbithole.Client)
+
+	user := d.Get("user").(string)
+	vhost := d.Get("vhost").(string)
+	permsList := d.Get("permissions").([]interface{})
+
+	for _, exchange := range permsList {
+
+		permsMap, ok := exchange.(map[string]interface{})
+		if !ok {
+			return fmt.Errorf("Unable to parse permissions")
+		}
+
+		if err := setTopicPermissionsIn(rmqc, vhost, user, permsMap); err != nil {
+			return err
+		}
+	}
+
+	id := fmt.Sprintf("%s@%s", user, vhost)
+	d.SetId(id)
+
+	return ReadTopicPermissions(d, meta)
+}
+
+func ReadTopicPermissions(d *schema.ResourceData, meta interface{}) error {
+	rmqc := meta.(*rabbithole.Client)
+
+	permissionId := strings.Split(d.Id(), "@")
+	if len(permissionId) < 2 {
+		return fmt.Errorf("Unable to determine Permission ID")
+	}
+
+	user := permissionId[0]
+	vhost := permissionId[1]
+
+	userPerms, err := rmqc.GetTopicPermissionsIn(vhost, user)
+	if err != nil {
+		return checkDeleted(d, err)
+	}
+
+	log.Printf("[DEBUG] RabbitMQ: Topic permission retrieved for %s: %#v", d.Id(), userPerms)
+
+	d.Set("user", userPerms[0].User)
+	d.Set("vhost", userPerms[0].Vhost)
+
+	perms := make([]map[string]interface{}, len(userPerms))
+	for i, perm := range userPerms {
+		p := make(map[string]interface{})
+		p["exchange"] = perm.Exchange
+		p["write"] = perm.Write
+		p["read"] = perm.Read
+		perms[i] = p
+	}
+	d.Set("permissions", perms)
+
+	return nil
+}
+
+func UpdateTopicPermissions(d *schema.ResourceData, meta interface{}) error {
+	rmqc := meta.(*rabbithole.Client)
+
+	permissionId := strings.Split(d.Id(), "@")
+	if len(permissionId) < 2 {
+		return fmt.Errorf("Unable to determine Permission ID")
+	}
+
+	user := permissionId[0]
+	vhost := permissionId[1]
+
+	if d.HasChange("permissions") {
+		_, newPerms := d.GetChange("permissions")
+
+		newPermsList := newPerms.([]interface{})
+		for _, exchange := range newPermsList {
```

Comment:
Unfortunately update is a bit more complex than that...

To highlight that, you just have to change the `testAccTopicPermissionsConfig_update` constant in the tests by adding  a second permissions in the same resource:

```golang
const testAccTopicPermissionsConfig_update = `
resource "rabbitmq_vhost" "test" {
    name = "test"
}

resource "rabbitmq_user" "test" {
    name = "mctest"
    password = "foobar"
    tags = ["administrator"]
}

resource "rabbitmq_topic_permissions" "test" {
    user = "${rabbitmq_user.test.name}"
    vhost = "${rabbitmq_vhost.test.name}"
    permissions {
        exchange = ".*"
        write = ".*"
        read = ""
    }

    permissions {
        exchange = "test"
	write = ".*"
        read = ".*"
    }
}`
```
Tests will fail.

The way you did it does not manage:
* Delete the removed permissions block
* `GetTopicPermissionsIn` does not seem to return permissions in the same order you created them (need to be checked), so there's alway a diff and `terraform apply` will try to update them endlessly.

If you keep `permissions` in this format, the first one is not so hard to fix, the second one a bit more complex. I'm not sure how to do that actually :thinking: but maybe you'll have an idea... You'll probably need, in the read function, to sort permissions in the same way that they exist in the state.

Another solution (but maybe less cool for the end user usage), is to create one resource by user/vhost/exchange. So the resource definition will look like (in a simplified view):
```
"user": String, ForceNew: true
"vhost": String, ForceNew: true
"exchange": String, ForceNew: true
"permissions": List, MaxItems: 1, {
    "read": String
    "write": Srtring
}, 
```

Reply:
I will have a :thinking:

Reply:
Switching to `schema.TypeSet` for `permissions`; which as well as being correct; fixes issues where the terraform order and RabbitMQ order differ.
Rabbit-hole currently only supports deleting all permissions for a vhost/user. So to do an update we delete all, then recreate the existing and new ones. This is not ideal as it could lead to removing permissions if you try and add an invalid one.

Reply:
Opened a PR upstream for just deleting the removed permissions: https://github.com/michaelklishin/rabbit-hole/pull/147

Reply:
> Opened a PR upstream for just deleting the removed permissions: michaelklishin/rabbit-hole#147

Thanks a lot :clap: :muscle:  , I was about to do the same.

> This is not ideal as it could lead to removing permissions if you try and add an invalid one.

Indeed!